### PR TITLE
GH-41706: [C++][Acero] Enhance asof_join to work in multi-threaded execution by sequencing input

### PR DESCRIPTION
**Pull request by mroz45**
### Rationale for this change
This is initial PR. I found that with specyfics parameters test fails.

### What changes are included in this PR?
In this PR I provoke fail of asof_join_node_test.

* GitHub Issue: #41706

### CODE REVIEW

**Review by westonpace — CHANGES_REQUESTED** [review]
I've let CI run, there are some formatting changes needed.  Otherwise, this seems straightforward enough.

The most significant change would probably be the breaking change that the asof join node now requires sequenced input.  However, I feel like this was always true in theory.

Perhaps @zanmato1984 @ZhangHuiGui @icexelloss or @rtpsw have thoughts?  I'm not certain who all is still using this node.

**Comment by westonpace** [comment]
This CI failure seems relevant:

```
FAILED: src/arrow/acero/CMakeFiles/arrow_acero_objlib.dir/asof_join_node.cc.o 
/usr/bin/ccache /usr/lib/ccache/clang++-14 -DADDRESS_SANITIZER -DARROW_ACERO_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_UBSAN -DJSON_DIAGNOSTICS=0 -DJSON_USE_IMPLICIT_CONVERSIONS=1 -I/build/cpp/src -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /build/cpp/opentelemetry_ep-install/include -Qunused-arguments -fcolor-diagnostics  -Wall -Wextra -Wdocumentation -DARROW_WARN_DOCUMENTATION -Wshorten-64-to-32 -Wno-missing-braces -Wno-unused-parameter -Wno-constant-logical-operand -Wno-return-stack-address -Wdate-time -Wno-unknown-warning-option -Wno-pass-failed -msse4.2  -fsanitize=address -DADDRESS_SANITIZER -fsanitize=undefined -fno-sanitize=alignment,vptr,function,float-divide-by-zero -fno-sanitize-recover=all -fsanitize-coverage=pc-table,inline-8bit-counters,edge,no-prune,trace-cmp,trace-div,trace-gep -fsanitize-blacklist=/arrow/cpp/build-support/sanitizer-disallowed-entries.txt -g -Werror -O0 -ggdb -g1 -fPIC   -fsanitize-coverage=pc-table,inline-8bit-counters,edge,no-prune,trace-cmp,trace-div,trace-gep -std=c++17 -MD -MT src/arrow/acero/CMakeFiles/arrow_acero_objlib.dir/asof_join_node.cc.o -MF src/arrow/acero/CMakeFiles/arrow_acero_objlib.dir/asof_join_node.cc.o.d -o src/arrow/acero/CMakeFiles/arrow_acero_objlib.dir/asof_join_node.cc.o -c /arrow/cpp/src/arrow/acero/asof_join_node.cc
In file included from /arrow/cpp/src/arrow/acero/asof_join_node.cc:18:
In file included from /arrow/cpp/src/arrow/acero/asof_join_node.h:20:
In file included from /arrow/cpp/src/arrow/acero/options.h:21:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/memory:76:
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:85:2: error: delete called on non-final 'arrow::acero::InputState' that has virtual functions but non-virtual destructor [-Werror,-Wdelete-non-abstract-non-virtual-dtor]
        delete __ptr;
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of member function 'std::default_delete<arrow::acero::InputState>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/arrow/cpp/src/arrow/acero/asof_join_node.cc:518:12: note: in instantiation of member function 'std::unique_ptr<arrow::acero::InputState>::~unique_ptr' requested here
    return std::make_unique<InputState>(index, tolerance, must_hash, may_rehash,
           ^
1 error generated.
```

**Comment by gitmodimo** [comment]
Let me pitch in. Disclaimer I am working @mroz45 on the same project using arrow.


```
                           -bool require_sequenced_output = false)
                           +bool require_sequenced_output = true)
```
> Changing this default would be a breaking change and I'm not certain it's warranted.

Without it python tests are failing.

> Should this only be `Ordering::Implicit` if `require_sequenced_output` is set?

It requires deeper and breaking changes which I think are necessary.
`require_sequenced_output` - means the source should give implicit ordering to produces batches ant therefore `require_sequenced_output` should be moved from `ScanNodeOptions` to `ScanOptions` to allow pass this option to `ScannerBuilder` which is used by python. But in fact I think there should be unified way to assert implicit ordering in all source nodes. Or maybe the _need_ for implicit ordering should propagate from nodes that need ordering (asof_join, fetch etc.) down the line to source nodes (and maybe fail if the source node cannot provide it). There are few related issues:
[no standardized sorting information](https://github.com/apache/arrow/issues/34451)
[add ordering information to exec batches](https://github.com/apache/arrow/issues/32991)
[Add AsofJoin Ordering Assertion](https://github.com/apache/arrow/issues/20353)

This [issue](https://github.com/apache/arrow/issues/27651) gave me the idea that implicit ordering should be asserted by default. And additional source node/additional option to assert no ordering - to enable some performance optimization for "don't care" ordering cases. This would fix those issues:
[asof_join node not working propertly](https://github.com/apache/arrow/issues/41706)
[order is unstable](https://github.com/apache/arrow/issues/15144)
[Preserve order when writing dataset](https://github.com/apache/arrow/issues/26818)
[ordering is weird](https://github.com/apache/arrow/issues/37542)
[dataset not preserving ordering](https://github.com/apache/arrow/issues/39030)
[scan node not asserting ordering](https://github.com/apache/arrow/issues/34698)

We are willing to contribute to fix ordering issue within acero but we have next to none experience in python/Cython. Also the size of the issue seems to grow with every little change. I think the ordering in Acero is a little bigger topic to discuss.

**Comment by gitmodimo** [comment]
The addition of ordering to python was more straightforward than I initially anticipated. Current version of pull request should fix at least some of above-mentioned python issues. Please review python changes carefully since it was our first contact with cython. 

PS. Can we also [Export SequencingQueue and SerialSequencingQueue
](https://github.com/gitmodimo/arrow/commit/03861ad1f0286fc4eee79532d0e19363b97a101d) This is useful in implementing custom nodes that require ordering. It seems reasonable to export it for custom nodes developers. Or maybe I should create separate pull request?

**Review by westonpace — APPROVED** [review]
Thanks for continuing to work on this and thank you for figuring out how to avoid the breaking change for now.

> Or maybe the need for implicit ordering should propagate from nodes that need ordering (asof_join, fetch etc.) down the line to source nodes (and maybe fail if the source node cannot provide it)

This seems reasonable but might be a bit complex.

> This https://github.com/apache/arrow/issues/27651 gave me the idea that implicit ordering should be asserted by default. And additional source node/additional option to assert no ordering - to enable some performance optimization for "don't care" ordering cases. This would fix those issues:

This seems logical to me.  Make the default safe but slow and add a potentially dangerous "I know what I'm doing" option to speed it up but still keep the responsibility on the user to figure out if that speedup is warranted.

I think it's probably pretty safe to change the defaults in things like `acero.py`.  My only concern would be changing defaults lower level in C++ which might change defaults in things like `dataset.py` which uses Acero and has a much wider user base than `acero.py`.  However, it may be that the usage of Acero from datasets already sets this parameter in all cases and doesn't rely on defaults (it's been too long since I've really dug in).  So if we feel that is the case then changing defaults lower down is probably fine too.

I'm +1 on this change provided CI gets to a good state.  Feel free to `@` me on this issue if you make any changes and need CI re-run.

**Comment by gitmodimo** [comment]
It looks like all checks passed. There are probably few issues (like [this](https://github.com/apache/arrow/issues/41706), [this maybe](https://github.com/apache/arrow/issues/34698) ) that should be verified and closed by this pull request. I think some of this work also overlaps with [GH-26818](https://github.com/apache/arrow/pull/44470)

**Comment by westonpace** [comment]
@gitmodimo thanks for the help getting this sorted out!

**Comment by gitmodimo** [comment]
> @gitmodimo thanks for the help getting this sorted out!

My pleasure. Me and @mroz45 have few more enhancements for upstreaming. We will post them soon.

File: cpp/src/arrow/acero/source_node.cc [hunk 1]
```diff
@@ -406,7 +407,7 @@ struct SchemaSourceNode : public SourceNode {
 struct RecordBatchReaderSourceNode : public SourceNode {
   RecordBatchReaderSourceNode(ExecPlan* plan, std::shared_ptr<Schema> schema,
                               arrow::AsyncGenerator<std::optional<ExecBatch>> generator)
-      : SourceNode(plan, schema, generator) {}
+      : SourceNode(plan, schema, generator, Ordering::Implicit()) {}
```
Comment (EnricoMi) [inline cpp/src/arrow/acero/source_node.cc:410]: Isn't this a bit too restrictive? This means all RecordBatch reader source **always** preserve the implicit order, even if if users do not care? Shouldn't this be configurable? What about other sources? Shouldn't the behaviour be the same for any source?

File: python/pyarrow/acero.py [hunk 2]
```diff
@@ -311,13 +313,18 @@ def _perform_join_asof(left_operand, left_on, left_by,
...
+            require_sequenced_output=True)
     else:
         left_source = Declaration(
             "table_source", TableSourceNodeOptions(left_operand),
         )
     if isinstance(right_operand, ds.Dataset):
-        right_source = _dataset_to_decl(right_operand, use_threads=use_threads)
+        right_source = _dataset_to_decl(
+            right_operand, use_threads=use_threads,
+            require_sequenced_output=True)
```
Comment (EnricoMi) [inline python/pyarrow/acero.py:327]: Do you really need the dataset to be sequenced? I think all you need is implicit ordering (batch indices) so you can sequence the batches at any point down the pipeline (by your asof join node).

File: cpp/src/arrow/dataset/scanner.cc [hunk 3]
```diff
@@ -1057,9 +1057,12 @@ Result<acero::ExecNode*> MakeScanNode(acero::ExecPlan* plan,
         batch->values.emplace_back(partial.record_batch.index);
         batch->values.emplace_back(partial.record_batch.last);
         batch->values.emplace_back(partial.fragment.value->ToString());
+        if (index != compute::kUnsequencedIndex) batch->index = index++;
         return batch;
       });
```
[the remark this answers is not in the document]
Comment (gitmodimo) [reply cpp/src/arrow/dataset/scanner.cc:1062]: It seems so. I first added indexing and then fixed ordering information propagation to source_node. Not sure it is possible to update already merged PR.

File: cpp/src/arrow/acero/source_node.cc (hunk 1 above)
Comment (gitmodimo) [reply cpp/src/arrow/acero/source_node.cc:410]: Probably option to specify ordering would be best. Or maybe automatically assign implicit ordering based on 
`batch.index!=compute::kUnsequencedIndex`?

File: cpp/src/arrow/dataset/scanner.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (EnricoMi) [reply cpp/src/arrow/dataset/scanner.cc:1062]: It's not possible, I'll draft a followup PR, if you don't mind.

File: python/pyarrow/acero.py (hunk 2 above)
Comment (gitmodimo) [reply python/pyarrow/acero.py:327]: > Do you really need the dataset to be sequenced? I think all you need is implicit ordering (batch indices) so you can sequence the batches at any point down the pipeline (by your asof join node).

True. But I don't think `require_sequenced_output` actually means batches should be produced by this node in order, but batches need to be sequenced (ie. have batch.index assigned). Current implementation _does_ produce batches in order as a side effect of indexing.

In context of your [GH-26818](https://github.com/apache/arrow/pull/44470)
Maybe the option `require_sequenced_output` should just be renamed to `implicit_ordering`?
What would be the purpose of `require_sequenced_output!=implicit_ordering`?

P.S. I think that the idea of "Dataset has implicit ordering" is not handled in arrow. There is no way to store ordering information in dataset and order is subjected to alphabetical ordering based on fragment filenames(which with default filename template is incorrect `part_1 -> part_10 -> part2 -> part_3` etc...).

File: cpp/src/arrow/acero/source_node.cc (hunk 1 above)
Comment (EnricoMi) [reply cpp/src/arrow/acero/source_node.cc:410]: Can you elaborate on why your asof_join enhancement requires specifically the `RecordBatchReaderSrouceNode` to have implicit ordering?

Comment (gitmodimo) [reply cpp/src/arrow/acero/source_node.cc:410]: It does not. I assumed that when user provides a generator,  it has some kind of implicot order.  Now I see it is not necessarily a case.

**Comment by EnricoMi** [comment]
@gitmodimo I have reworked some of this PR's changes in #44616 and added some of my code from #44470. Please have a look.

**Pull request merged**